frontend/app/components/work-packages/work-package-cache.service.ts

```diff
@@ -30,51 +30,55 @@
 import {opWorkPackagesModule} from "../../angular-modules";
 import {WorkPackageResource} from "../api/api-v3/hal-resources/work-package-resource.service";
 import {ApiWorkPackagesService} from "../api/api-work-packages/api-work-packages.service";
+import {states} from "../../states";
+import {State} from "../../helpers/reactive-fassade";
 import IScope = angular.IScope;
 
 
 export class WorkPackageCacheService {
 
-  private workPackageCache: {[id: number]: WorkPackageResource} = {};
-
-  private workPackagesSubject = new Rx.ReplaySubject<{[id: number]: WorkPackageResource}>(1);
-
   private newWorkPackageCreatedSubject = new Rx.Subject<WorkPackageResource>();
 
   /*@ngInject*/
-  constructor(private $rootScope: IScope, private apiWorkPackages: ApiWorkPackagesService) {
+  constructor(private apiWorkPackages: ApiWorkPackagesService) {
   }
 
   newWorkPackageCreated(wp: WorkPackageResource) {
     this.newWorkPackageCreatedSubject.onNext(wp);
   }
 
   updateWorkPackage(wp: WorkPackageResource) {
-    this.updateWorkPackageList([wp]);
+    states.workPackages.put(wp.id.toString(), wp);
   }
 
   updateWorkPackageList(list: WorkPackageResource[]) {
-    for (const wp of list) {
-      var cached = this.workPackageCache[wp.id];
-      if (cached && cached.dirty) {
-        this.workPackageCache[wp.id] = cached;
-      } else {
-        this.workPackageCache[wp.id] = wp;
+    for (var wp of list) {
+
+      // var cached = this.workPackageCache[wp.id];
+      // if (cached && cached.dirty) {
+      //   this.workPackageCache[wp.id] = cached;
+      // } else {
+      //   this.workPackageCache[wp.id] = wp;
+      // }
+      const wpState = states.workPackages.get(wp.id.toString());
+      if (wpState.hasValue() && wpState.getLastValue().dirty) {
+        return;
       }
+
+      states.workPackages.put(wp.id.toString(), wp);
     }
-    this.workPackagesSubject.onNext(this.workPackageCache);
   }
 
-  loadWorkPackage(workPackageId: number, forceUpdate = false): Rx.Observable<WorkPackageResource> {
-    if (forceUpdate || this.workPackageCache[workPackageId] === undefined) {
-      this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate).then(wp => {
-        this.updateWorkPackage(wp);
-      });
+  loadWorkPackage(workPackageId: number, forceUpdate = false): State<WorkPackageResource> {
+    const state = states.workPackages.get(workPackageId.toString());
+    if (forceUpdate) {
+      state.clear();
     }
 
-    return this.workPackagesSubject
-      .map(cache => cache[workPackageId])
-      .filter(wp => wp !== undefined);
+    state.putFromPromiseIfPristine(
+      () => this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate));
+
+    return state;
```

Comment:
state.isPristine() will be true after state.clear() because...

- clear() sets `null` as the state
- setState sets `timestampOfLastValue` to `-1` due to the null value
- isPristine() therefore returns true